File: backend/Models/WatcherEvent.cs
namespace GSSystemAnalyzer.Models
{
    public enum WatcherChangeKind
    {
        Created,
        Modified,
        Deleted,
        Renamed,
        Overflow
    }

    public record WatcherEvent
    {
        public WatcherEvent(DateTimeOffset timestamp, WatcherChangeKind kind, string path, string? oldPath, bool isDirectory, int occurrences)
        {
            Timestamp = timestamp;
            Kind = kind;
            Path = path;
            OldPath = oldPath;
            IsDirectory = isDirectory;
            Occurrences = occurrences;
        }

        public DateTimeOffset Timestamp { get; init; }
        public WatcherChangeKind Kind { get; init; }
        public string Path { get; init; }
        public string? OldPath { get; init; }
        public bool IsDirectory { get; init; }
        public int Occurrences { get; set; }
    }
}

File: frontend/gs_analyzer_ui/lib/models/watcher_event.dart
enum WatcherChangeKind {
  created,
  modified,
  deleted,
  renamed,
  overflow;

  static WatcherChangeKind fromInt(int value) {
    if (value >= 0 && value < values.length) {
      return values[value];
    }
    return modified;
  }
}

class WatcherEvent {
  final DateTime timestamp;
  final WatcherChangeKind kind;
  final String path;
  final String? oldPath;
  final bool isDirectory;
  final int occurrences;

  WatcherEvent({
    required this.timestamp,
    required this.kind,
    required this.path,
    this.oldPath,
    required this.isDirectory,
    required this.occurrences,
  });

  factory WatcherEvent.fromJson(Map<String, dynamic> json) {
    // If backend serializes as int or string, we should handle both securely.
    WatcherChangeKind kind;
    if (json['kind'] is int) {
      kind = WatcherChangeKind.fromInt(json['kind']);
    } else {
      kind = WatcherChangeKind.values.firstWhere(
        (e) => e.name.toLowerCase() == json['kind'].toString().toLowerCase(),
        orElse: () => WatcherChangeKind.modified,
      );
    }

    return WatcherEvent(
      timestamp: DateTime.parse(json['timestamp']),
      kind: kind,
      path: json['path'],
      oldPath: json['oldPath'],
      isDirectory: json['isDirectory'] ?? false,
      occurrences: json['occurrences'] ?? 1,
    );
  }
}
